src/test/java/com/amplitude/api/DeviceInfoTest.java

```diff
@@ -304,4 +312,90 @@ public void testDontUseAdvertisingIdAsDeviceId() {
         assertNotEquals(advertisingId, client.getDeviceId());
         assertTrue(client.getDeviceId().endsWith("R"));
     }
+
+    @Test
+    public void testAppSetId() {
+        PowerMockito.mockStatic(AdvertisingIdClient.class);
+        String advertisingId = "advertisingId";
+        AdvertisingIdClient.Info info = new AdvertisingIdClient.Info(
+                advertisingId,
+                true
+        );
+        try {
+            Mockito.when(AdvertisingIdClient.getAdvertisingIdInfo(context)).thenReturn(info);
+        } catch (Exception e) {
+            fail(e.toString());
+        }
+
+        PowerMockito.mockStatic(UUID.class);
+        String mockUUIDString = "a74d2fb3-7c64-4d60-b0ca-b633aa7a4dbc";
+        UUID mockUUID = UUID.fromString(mockUUIDString);
+        try {
+            Mockito.when(UUID.randomUUID()).thenReturn(mockUUID);
+        } catch (Exception e) {
+            fail(e.toString());
+        }
+
+        String mockAppSetId = "5a8f0fd1-31a9-4a1f-bfad-cd5439ce533b";
+        PowerMockito.stub(PowerMockito.method(DeviceInfo.class, "getAppSetId"))
+                .toReturn(mockAppSetId);
+
+        Robolectric.getForegroundThreadScheduler().advanceTo(1);
+
+        AmplitudeClient client = Amplitude.getInstance("UseAppSetId");
+        client.useAppSetIdForDeviceId();
+        client.initialize(context, "1cc2c1978ebab0f6451112a8f5df4f4e");
+        ShadowLooper looper = Shadows.shadowOf(client.logThread.getLooper());
+        looper.runToEndOfTasks();
+
+        assertNotEquals(advertisingId, client.getDeviceId());
+        assertNotEquals(mockUUIDString, client.getDeviceId());
+        assertTrue(client.getDeviceId().endsWith("S"));
+    }
+
+    @Test
+    public void testDoNotSendAppSetId() {
+        String mockAppSetId = "5a8f0fd1-31a9-4a1f-bfad-cd5439ce533b";
+        PowerMockito.stub(PowerMockito.method(DeviceInfo.class, "getAppSetId"))
```

Comment:
This line seems the place causing issues and we have to import the lib? What if when we initialize we create a new method `collectDeviceInfo` from our client, then we can mock this method on amplitude interface to return a mocked DeviceInfo object, then we can just use Mockito instead of PowerMock in this test case. And if that works, we don't need to import the lib.

Reply:
That's an interesting idea here. I'll try to implement on my end

Reply:
I considered doing this, but I didn't want to touch any private functions or change the encapsulation. My understanding is that if a method is private, it shouldn't be involved as the main focus of a test. 

I'm totally fine with just importing `json...` as a `testImplementation` dependency. Our code does touch the `android json` library, and from stackoverflow, mockito tests of any kind should not rely on built-in Android libraries. It's their strange rule, you can't involve others' code in your test where you are mocking your own code.

Reply:
I guess it might be fine to leave it like this, just for best practice we should try to avoid that as possible. Also as mocking happening on DeviceInfo, and there is no JSONObject usage inside that class, it is a bit confusing to adding in JSONObject as part of test dependency. For better practice, we may want to avoid extra dependency and much PowerMock usage.